File: String/165. Compare Version Numbers.cpp
class Solution {
public:
    vector<int> getValue(string s)
    {
        vector<int>ret;
        string x;
        for(char c:s)
        {
            if(c=='.')
            {
                ret.push_back(stoi(x));
                x.clear();
            }
            else x += c;
        }
        ret.push_back(stoi(x));
        return ret;
    }
    int compareVersion(string version1, string version2) 
    {
        vector<int> v1 = getValue(version1), v2 = getValue(version2);
        while(v1.size() and v1.back()==0) v1.pop_back();
        while(v2.size() and v2.back()==0) v2.pop_back();
        if(v1 == v2) return 0;

        int n = min(v1.size(), v2.size());
        for(int i=0;i<n;i++)
        {
            if(v1[i] > v2[i]) return 1;
            else if(v1[i] < v2[i]) return -1;
        }
        return v1.size() > v2.size() ? 1 : -1;
    }
};

File: String/165. Compare Version Numbers.py
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        def spliteVersion(s, code=""):
            arr = []
            for ch in s:
                if ch == '.':
                    arr.append(int(code))
                    code = ""
                else:
                    code += ch
            while arr and arr[-1] == 0:
                arr.pop()
            return arr

        v1_arr = spliteVersion(version1 + ".")
        v2_arr = spliteVersion(version2 + ".")
        if v1_arr == v2_arr:
            return 0

        n = min(len(v1_arr), len(v2_arr))
        for i in range(n):
            if v1_arr[i] < v2_arr[i]:
                return -1
            if v1_arr[i] > v2_arr[i]:
                return 1

        return 1 if len(v1_arr) > len(v2_arr) else -1
    


class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        def spliteVersion(s, code=""):
            arr = []
            for ch in s:
                if ch == '.':
                    arr.append(code)
                    code = ""
                else:
                    code += ch
            while arr and int(arr[-1]) == 0:
                arr.pop()
            return arr

        v1_arr = spliteVersion(version1 + ".")
        v2_arr = spliteVersion(version2 + ".")

        n = min(len(v1_arr), len(v2_arr))
        for i in range(n):
            if int(v1_arr[i]) < int(v2_arr[i]):
                return -1
            if int(v1_arr[i]) > int(v2_arr[i]):
                return 1

        v1_sum =  sum([int(s) for s in v1_arr])
        v2_sum =  sum([int(s) for s in v2_arr])

        if v1_sum < v2_sum:
            return -1
        if v1_sum > v2_sum:
            return 1
        return 0
    